File: app/src/main/java/ru/art2000/extensions/views/FloatingActionButtonScrollingBehaviour.java
package ru.art2000.extensions.views;

import android.animation.Animator;
import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;

import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.view.ViewCompat;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

public class FloatingActionButtonScrollingBehaviour extends FloatingActionButton.Behavior {

    private static final int SHOWN = 0;
    private static final int HIDDEN = 1;
    private static final int ANIMATING_DOWN = 2;
    private static final int ANIMATING_UP = 3;
    private static final long ANIMATION_DURATION = 100;
    private static final int SCROLL_OFFSET = 1;
    private int state = SHOWN;

    private boolean isListenerAdded = false;

    public FloatingActionButtonScrollingBehaviour(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean onStartNestedScroll(@NonNull CoordinatorLayout coordinatorLayout,
                                       @NonNull FloatingActionButton child,
                                       @NonNull View directTargetChild,
                                       @NonNull View target,
                                       int axes,
                                       int type) {
        if (!isListenerAdded) {
            isListenerAdded = true;
            child.addOnShowAnimationListener(new Animator.AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animator) {
                    child.setTranslationY(0);
                    state = SHOWN;
                }

                @Override
                public void onAnimationEnd(Animator animator) {

                }

                @Override
                public void onAnimationCancel(Animator animator) {

                }

                @Override
                public void onAnimationRepeat(Animator animator) {

                }
            });
        }
        return axes == ViewCompat.SCROLL_AXIS_VERTICAL;
    }

    @Override
    public void onNestedScroll(@NonNull CoordinatorLayout coordinatorLayout,
                               @NonNull FloatingActionButton child,
                               @NonNull View target,
                               int dxConsumed, int dyConsumed,
                               int dxUnconsumed, int dyUnconsumed,
                               int type,
                               @NonNull int[] consumed) {
        setupAnimation(child, dyConsumed);
        super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed,
                dxUnconsumed, dyUnconsumed, type, consumed);
    }

    @Override
    public void onNestedPreScroll(@NonNull CoordinatorLayout coordinatorLayout,
                                  @NonNull FloatingActionButton child,
                                  @NonNull View target,
                                  int dx, int dy,
                                  @NonNull int[] consumed,
                                  int type) {
        setupAnimation(child, dy);
        super.onNestedPreScroll(coordinatorLayout, child, target, dx, dy, consumed, type);
    }

    private void setupAnimation(FloatingActionButton child, int dy) {
        if (dy > SCROLL_OFFSET && state == SHOWN) {
            state = ANIMATING_DOWN;
            CoordinatorLayout.MarginLayoutParams layoutParams =
                    (CoordinatorLayout.MarginLayoutParams) child.getLayoutParams();
            child.animate()
                    .translationY(child.getHeight() + layoutParams.bottomMargin)
                    .setInterpolator(new LinearInterpolator())
                    .setDuration(ANIMATION_DURATION)
                    .setListener(new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animator) {

                        }

                        @Override
                        public void onAnimationEnd(Animator animator) {
                            state = HIDDEN;
                        }

                        @Override
                        public void onAnimationCancel(Animator animator) {

                        }

                        @Override
                        public void onAnimationRepeat(Animator animator) {

                        }
                    }).start();
        } else if (dy < -SCROLL_OFFSET && state == HIDDEN) {
            state = ANIMATING_UP;
            child.animate()
                    .translationY(0)
                    .setInterpolator(new LinearInterpolator())
                    .setDuration(ANIMATION_DURATION)
                    .setListener(new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animator) {

                        }

                        @Override
                        public void onAnimationEnd(Animator animator) {
                            state = SHOWN;
                        }

                        @Override
                        public void onAnimationCancel(Animator animator) {

                        }

                        @Override
                        public void onAnimationRepeat(Animator animator) {

                        }
                    }).start();
        }
    }
}

File: app/src/main/java/ru/art2000/extensions/views/FloatingActionButtonScrollingBehaviour.kt
package ru.art2000.extensions.views

import android.animation.Animator
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import android.view.animation.LinearInterpolator
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.ViewCompat

class FloatingActionButtonScrollingBehaviour(context: Context, attrs: AttributeSet?) :
    CoordinatorLayout.Behavior<View>(context, attrs) {

    private var state = SHOWN

    override fun onStartNestedScroll(
        coordinatorLayout: CoordinatorLayout,
        child: View,
        directTargetChild: View,
        target: View,
        axes: Int,
        type: Int
    ): Boolean {
        return axes == ViewCompat.SCROLL_AXIS_VERTICAL
    }

    override fun onNestedScroll(
        coordinatorLayout: CoordinatorLayout,
        child: View,
        target: View,
        dxConsumed: Int, dyConsumed: Int,
        dxUnconsumed: Int, dyUnconsumed: Int,
        type: Int,
        consumed: IntArray
    ) {
        setupAnimation(child, dyConsumed)
        super.onNestedScroll(
            coordinatorLayout, child, target, dxConsumed, dyConsumed,
            dxUnconsumed, dyUnconsumed, type, consumed
        )
    }

    override fun onNestedPreScroll(
        coordinatorLayout: CoordinatorLayout,
        child: View,
        target: View,
        dx: Int, dy: Int,
        consumed: IntArray,
        type: Int
    ) {
        setupAnimation(child, dy)
        super.onNestedPreScroll(coordinatorLayout, child, target, dx, dy, consumed, type)
    }

    private fun setupAnimation(child: View, dy: Int) {
        if (dy > SCROLL_OFFSET && state == SHOWN) {
            state = ANIMATING_DOWN
            val layoutParams = child.layoutParams as MarginLayoutParams
            child.animate()
                .translationY((child.height + layoutParams.bottomMargin + 8).toFloat())
                .setInterpolator(LinearInterpolator())
                .setDuration(ANIMATION_DURATION)
                .setListener(object : Animator.AnimatorListener {
                    override fun onAnimationStart(animator: Animator) {}
                    override fun onAnimationEnd(animator: Animator) {
                        state = HIDDEN
                    }

                    override fun onAnimationCancel(animator: Animator) {}
                    override fun onAnimationRepeat(animator: Animator) {}
                }).start()
        } else if (dy < -SCROLL_OFFSET && state == HIDDEN) {
            state = ANIMATING_UP
            child.animate()
                .translationY(0f)
                .setInterpolator(LinearInterpolator())
                .setDuration(ANIMATION_DURATION)
                .setListener(object : Animator.AnimatorListener {
                    override fun onAnimationStart(animator: Animator) {}
                    override fun onAnimationEnd(animator: Animator) {
                        state = SHOWN
                    }

                    override fun onAnimationCancel(animator: Animator) {}
                    override fun onAnimationRepeat(animator: Animator) {}
                }).start()
        }
    }

    companion object {
        private const val SHOWN = 0
        private const val HIDDEN = 1
        private const val ANIMATING_DOWN = 2
        private const val ANIMATING_UP = 3
        private const val ANIMATION_DURATION: Long = 100
        private const val SCROLL_OFFSET = 1
    }
}